### PR TITLE
refactor(payments): add udf field and remove refactor metadata

### PR DESCRIPTION
**Pull request by Abhicodes-crypto**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

In this PR , 
`routing_parameters` and `data`  are removed from metadata.
`udf` field was added in `PaymentsRequest`

Backwards compatibility for order_details in metadata is maintained but will be deprecated in the future.

### Additional Changes

- [x] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

I tested everything and it worked fine. 
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -1641,6 +1641,22 @@ pub struct RedirectResponse {
...
 }
 
+#[derive(
+    Default,
+    Debug,
+    serde::Deserialize,
+    serde::Serialize,
+    Clone,
+    ToSchema,
+    router_derive::PolymorphicSchema,
```
Comment (lsampras) [inline crates/api_models/src/payments.rs:1651]: Do we need `PolymorphicSchema` here?

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -239,16 +244,15 @@ pub struct PaymentsRequest {
...
-
     /// Business sub label for the payment
     pub business_sub_label: Option<String>,
 
     /// If enabled payment can be retried from the client side until the payment is successful or payment expires or the attempts(configured by the merchant) for payment are exhausted.
     #[serde(default)]
     pub manual_retry: bool,
+
+    /// metadata that will be deprecated. For now only order details can be passed here.
+    pub metadata: Option<MetadataExternal>,
```
Comment (lsampras) [inline crates/api_models/src/payments.rs:255]: will need to add schema here?

probably can specify in the schema that this field is gonna be deprecated

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 3]
```diff
@@ -103,6 +103,8 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsRequest> for Pa
             .map(ForeignInto::foreign_into)
             .or(payment_intent.setup_future_usage);
 
+        payment_intent.udf = payment_intent.udf.or(request.udf.clone());
```
Comment (lsampras) [inline crates/router/src/core/payments/operations/payment_confirm.rs:106]: If the merchant hasn't specified udf when he created intent,
we shouldn't be picking one up from the confirm request as well right?

File: crates/api_models/src/payments.rs [hunk 4]
```diff
@@ -1355,12 +1360,14 @@ pub struct PaymentsResponse {
...
-    pub allowed_payment_method_types: Option<Vec<api_enums::PaymentMethodType>>,
-
     /// ephemeral_key for the customer_id mentioned
     pub ephemeral_key: Option<EphemeralKeyCreateResponse>,
+
+    /// the metadata that will be deprecated
+    pub metadata: Option<MetadataExternal>,
+
+    /// user defined values
+    pub udf: Option<pii::SecretSerdeValue>,
```
Comment (lsampras) [inline crates/api_models/src/payments.rs:1370]: Why are these fields not included in API schema

File: crates/storage_models/src/payment_intent.rs [hunk 5]
```diff
@@ -122,6 +124,7 @@ pub enum PaymentIntentUpdate {
         business_label: Option<String>,
         order_details: Option<Vec<pii::SecretSerdeValue>>,
         metadata: Option<pii::SecretSerdeValue>,
+        udf: Option<pii::SecretSerdeValue>,
```
Comment (lsampras) [inline crates/storage_models/src/payment_intent.rs:127]: we could remove this if udf is not intented to be updated...

**Review by lsampras — COMMENTED** [review]
I don't understand the motivation behind the metada related change...

> From a users perspective there is no field metadata. He can populate fields inside metadata by adding the same fields in the payment request directly.

1. since we don't add metadata in the schema he doesn't know about it...
2. But why would he even add fields like `order_details` if they are also not present in our API schema...
3. If we do decide to make such fields public why not restrict them to using metadata...
4. In general we should keep the struct format & API request json similar in structure so that we can easily correlate the request when looking at logs...

Unless there's a compelling need to flatten the metadata I'd prefer it to remain the same...

File: crates/api_models/src/payments.rs (hunk 4 above)
Comment (Abhicodes-crypto) [reply crates/api_models/src/payments.rs:1370]: `metadata` would be deprecated so we don't want it in API schema. Will add udf.

File: crates/router/src/core/payments/operations/payment_confirm.rs (hunk 3 above)
Comment (Abhicodes-crypto) [reply crates/router/src/core/payments/operations/payment_confirm.rs:106]: I feel the same But the task was to add udf to PaymentsRequest and whichever API's using it.

File: crates/api_models/src/payments.rs [hunk 6]
```diff
@@ -184,7 +184,12 @@ pub struct PaymentsRequest {
     pub statement_descriptor_suffix: Option<String>,
 
     /// You can specify up to 50 keys, with key names up to 40 characters long and values up to 500 characters long. Metadata is useful for storing additional, structured information on an object.
-    pub metadata: Option<Metadata>,
+    #[serde(flatten)]
+    pub metadata_internal: Option<Metadata>,
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:188]: Follow the specs to name the fields and structs.

This field is metadata only.

File: crates/api_models/src/payments.rs [hunk 7]
```diff
@@ -239,16 +244,15 @@ pub struct PaymentsRequest {
...
-    pub allowed_payment_method_types: Option<Vec<api_enums::PaymentMethodType>>,
-
     /// Business sub label for the payment
     pub business_sub_label: Option<String>,
 
     /// If enabled payment can be retried from the client side until the payment is successful or payment expires or the attempts(configured by the merchant) for payment are exhausted.
     #[serde(default)]
     pub manual_retry: bool,
+
+    /// metadata that will be deprecated. For now only order details can be passed here.
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:254]: Remove this comment and add the valid comment, this field is not going to deprecated.

For naming convention refer the spec sheet

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/storage_models/src/payment_intent.rs (hunk 5 above)
Comment (Abhicodes-crypto) [reply crates/storage_models/src/payment_intent.rs:127]: It is being updated in payment update and confirm call

File: crates/api_models/src/payments.rs [hunk 8]
```diff
@@ -281,6 +281,10 @@ pub struct PaymentsRequest {
     /// If enabled payment can be retried from the client side until the payment is successful or payment expires or the attempts(configured by the merchant) for payment are exhausted.
     #[serde(default)]
     pub manual_retry: bool,
+
+    /// Any user defined fields can be passed here.
+    #[schema(value_type = Object, example = r#"{ "city": "NY", "unit": "245" }"#)]
+    pub udf: Option<pii::SecretSerdeValue>,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:287]:
```suggestion
    #[schema(value_type = Option<Object>, example = r#"{ "city": "NY", "unit": "245" }"#)]
    pub udf: Option<pii::SecretSerdeValue>,
```

File: crates/api_models/src/payments.rs [hunk 9]
```diff
@@ -1377,6 +1381,10 @@ pub struct PaymentsResponse {
 
     /// ephemeral_key for the customer_id mentioned
     pub ephemeral_key: Option<EphemeralKeyCreateResponse>,
+
+    /// Any user defined fields can be passed here.
+    #[schema(value_type = Object, example = r#"{ "city": "NY", "unit": "245" }"#)]
+    pub udf: Option<pii::SecretSerdeValue>,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1387]:
```suggestion
    #[schema(value_type = Option<Object>, example = r#"{ "city": "NY", "unit": "245" }"#)]
    pub udf: Option<pii::SecretSerdeValue>,
```

File: crates/api_models/src/payments.rs [hunk 10]
```diff
@@ -1636,12 +1644,7 @@ pub struct OrderDetails {
 pub struct Metadata {
     /// Information about the product and quantity for specific connectors. (e.g. Klarna)
     pub order_details: Option<OrderDetails>,
-    /// Information used for routing
-    pub routing_parameters: Option<HashMap<String, String>>,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1640]: Why was this field removed, is it no longer required? If it is required, then how are we accepting it?

Comment (Abhicodes-crypto) [reply crates/api_models/src/payments.rs:1640]: Yeah I have confirmed with @bernard-eugine and @vspecky . It can be removed.

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**